src/core/search/compressed_list.cc

```diff
@@ -0,0 +1,210 @@
+#include "core/search/compressed_list.h"
+
+#include <array>
+
+#include "base/logging.h"
+
+namespace dfly::search {
+
+using namespace std;
+
+CompressedList::Iterator::Iterator(const CompressedList& list) : stash_{}, diffs_{list.diffs_} {
+  ReadNext();
+}
+
+CompressedList::Iterator::Iterator() : stash_{}, diffs_{nullptr, 0} {
+}
+
+uint32_t CompressedList::Iterator::operator*() const {
+  DCHECK(stash_);
+  return *stash_;
+}
+
+CompressedList::Iterator& CompressedList::Iterator::operator++() {
+  ReadNext();
+  return *this;
+}
+
+bool operator==(const CompressedList::Iterator& l, const CompressedList::Iterator& r) {
+  return l.diffs_.data() == r.diffs_.data() && l.diffs_.size() == r.diffs_.size();
+}
+
+bool operator!=(const CompressedList::Iterator& l, const CompressedList::Iterator& r) {
+  return !(l == r);
+}
+
+void CompressedList::Iterator::ReadNext() {
+  if (diffs_.empty()) {
+    diffs_ = {nullptr, 0};
+    return;
```

Comment:
Should you also assign `nullopt` to `stash_` here?

Reply:
Theoretically it doesn't matter because reading from an iterator that is equal to end should be forbidden. But I'll add it for better safety